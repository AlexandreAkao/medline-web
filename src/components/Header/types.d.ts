interface IHeaderProps {
  icon: string;
  iconAlt: string;
  isAuthenticated: boolean;
  children: React.ReactNode;
}
