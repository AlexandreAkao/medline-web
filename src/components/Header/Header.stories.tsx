import { ComponentStory, ComponentMeta } from '@storybook/react';
import { FaSignOutAlt } from 'react-icons/fa';

import Header from 'components/Header';
import HeaderItem from 'components/HeaderItem';
import colors from 'styles/colors';

export default {
  title: 'Header',
  args: {
    isAuthenticated: false,
    items: [],
    iconAlt: 'icon alt',
  },
  decorators: [
    Story => (
      <div
        style={{
          background: colors.primary.normal,
          width: '100%',
          height: '100%',
        }}
      >
        <Story />
      </div>
    ),
  ],
  component: Header,
} as ComponentMeta<typeof Header>;

const Template: ComponentStory<typeof Header> = args => <Header {...args} />;

export const Authenticated = Template.bind({});
Authenticated.args = {
  icon: 'https://cdn-icons-png.flaticon.com/512/5277/5277459.png',
  isAuthenticated: true,
  children: (
    <>
      <HeaderItem as="link">Editar Perfil</HeaderItem>
      <HeaderItem as="link">Atestados/Receitas</HeaderItem>
      <HeaderItem as="link">Solicitação</HeaderItem>
      <HeaderItem as="link">
        Olá, user <FaSignOutAlt />
      </HeaderItem>
    </>
  ),
};

export const WithoutAuthenticated = Template.bind({});
WithoutAuthenticated.args = {
  icon: 'https://cdn-icons-png.flaticon.com/512/5277/5277459.png',
  isAuthenticated: false,
  children: (
    <>
      <HeaderItem as="button">Cadastro</HeaderItem>
      <HeaderItem as="link">Entrar</HeaderItem>
    </>
  ),
};
