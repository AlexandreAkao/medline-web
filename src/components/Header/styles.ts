import styled from 'styled-components';

import { HeaderItemContainer } from 'components/HeaderItem/styles';
import colors from 'styles/colors';
import zIndexPriority from 'styles/zIndexPriority';
import ScreenSizes from 'styles/screenSizes';

export const Nav = styled.header<Pick<IHeaderProps, 'isAuthenticated'>>`
  width: 100%;
  height: 80px;
  box-shadow: ${({ isAuthenticated }) =>
    isAuthenticated ? '0px 4px 10px 1px rgba(0, 0, 0, 0.25);' : null};
  z-index: ${zIndexPriority.VERY_HIGH};
  background-color: ${colors.primary.normal};
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: 0 30px;

  @media (max-width: ${ScreenSizes.medium}) {
    box-shadow: none;
  }
`;

export const LogoImg = styled.img`
  cursor: pointer;
  height: 65%;
`;

export const NavItemsContainer = styled.nav`
  display: flex;
  justify-content: center;
  align-items: center;

  ${HeaderItemContainer} {
    padding: 0 16px;
  }
`;
