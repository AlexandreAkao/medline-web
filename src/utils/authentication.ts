import { isEmpty } from 'ramda';

export const getUserInfo = () => {
  const user = JSON.parse(localStorage.getItem('user') ?? '{}');

  return isEmpty(user) ? null : (user as IUser);
};

export const geToken = () => localStorage.getItem('token');
