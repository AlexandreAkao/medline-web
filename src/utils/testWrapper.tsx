import { render } from '@testing-library/react';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';

export const renderWithRouter = (component: React.ReactNode) =>
  render(<MemoryRouter>{component}</MemoryRouter>);
