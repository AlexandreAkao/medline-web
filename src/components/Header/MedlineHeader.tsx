import { useMemo, useCallback } from 'react';
import { useNavigate } from 'react-router-dom';
import { FaRegUser, FaReceipt, FaPlay, FaUserEdit } from 'react-icons/fa';
import { MdExitToApp } from 'react-icons/md';
import { useMediaQuery } from 'react-responsive';

import Header from 'components/Header';
import { useAuth } from 'hooks/useAuth';
import MedlineLogo from 'assets/images/medline-logo.svg';
import HeaderItem from 'components/HeaderItem';
import OverflowMenu from 'components/OverflowMenu';
import ScreenSizes from 'styles/screenSizes';
import {
  HeaderMenuBurgerContainer,
  HeaderMenuBurgerTitle,
  OverflowMenuNavigation,
  OverflowMenuNavigationItem,
  OverflowMenuTitle,
} from 'components/Header/MedlineHeader.style';
import colors from 'styles/colors';

function MedlineHeader() {
  const navigate = useNavigate();
  const { signed, user, handleLogout } = useAuth();
  const isTablet = useMediaQuery({ query: `(max-width: ${ScreenSizes.medium})` });

  const handleClickRegister = useCallback(() => {
    navigate('/register');
  }, [navigate]);

  const navigationButtons = useMemo(() => {
    return signed && user ? (
      <>
        <HeaderItem as="link">Atestados/Receitas</HeaderItem>
        <HeaderItem as="link">Solicitação</HeaderItem>
        <HeaderItem as="link">
          <OverflowMenu render={<FaRegUser size={20} />}>
            <OverflowMenuTitle>Olá, {user.name}</OverflowMenuTitle>
            <OverflowMenuNavigation>
              <OverflowMenuNavigationItem>
                <FaUserEdit color={colors.grey.normal} fontSize={25} />
                <span>Editar Perfil</span>
              </OverflowMenuNavigationItem>
              <OverflowMenuNavigationItem onClick={handleLogout}>
                <MdExitToApp color={colors.grey.normal} fontSize={25} />
                <span>Sair</span>
              </OverflowMenuNavigationItem>
            </OverflowMenuNavigation>
          </OverflowMenu>
        </HeaderItem>
      </>
    ) : (
      <>
        <HeaderItem as="button" onClick={handleClickRegister}>
          Cadastro
        </HeaderItem>
        <HeaderItem as="link" href="/login">
          Entrar
        </HeaderItem>
      </>
    );
  }, [handleClickRegister, handleLogout, signed, user]);

  const navigationButtonsBurger = useMemo(() => {
    return signed && user ? (
      <HeaderMenuBurgerContainer>
        <HeaderMenuBurgerTitle>Olá, {user.name}</HeaderMenuBurgerTitle>
        <HeaderItem>
          <FaReceipt />
          <span>Atestados/Receitas</span>
        </HeaderItem>
        <HeaderItem>
          <FaPlay />
          <span>Solicitação</span>
        </HeaderItem>
        <HeaderItem>
          <FaUserEdit />
          <span>Editar perfil</span>
        </HeaderItem>
        <HeaderItem onClick={handleLogout}>
          <MdExitToApp />
          <span>Sair</span>
        </HeaderItem>
      </HeaderMenuBurgerContainer>
    ) : (
      <>
        <HeaderItem as="link" href="/register">
          Cadastro
        </HeaderItem>
        <HeaderItem as="link" href="/login">
          Entrar
        </HeaderItem>
      </>
    );
  }, [handleLogout, signed, user]);

  return (
    <Header icon={MedlineLogo} iconAlt="Medline logo" isAuthenticated={signed}>
      {isTablet ? navigationButtonsBurger : navigationButtons}
    </Header>
  );
}

export default MedlineHeader;
