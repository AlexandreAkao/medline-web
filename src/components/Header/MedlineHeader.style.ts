import styled from 'styled-components';

import colors from 'styles/colors';
import { HeaderItemContainer } from 'components/HeaderItem/styles';

export const HeaderMenuBurgerTitle = styled.h1`
  font-size: 2rem;
  color: ${colors.white.normal};
  border-bottom: 4px solid ${colors.black.soft};
  padding-bottom: 8px;
  margin-bottom: 2rem;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
`;

export const HeaderMenuBurgerContainer = styled.div`
  ${HeaderItemContainer} {
    a,
    button {
      justify-content: flex-start;
      font-size: 1.5rem;
      svg {
        margin-right: 3rem;
      }
    }
  }
`;

export const OverflowMenuTitle = styled.h1`
  font-size: 1.6rem;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
  padding: 1.6rem;
  color: ${colors.black.soft};
`;

export const OverflowMenuNavigation = styled.nav`
  border-top: 2px solid ${colors.black.soft};
  display: flex;
  flex-direction: column;
  padding: 1.6rem;

  ${HeaderItemContainer} {
    padding: 1rem;
    a {
      padding: 0;
      color: ${colors.grey.normal};
      justify-content: flex-start;

      :hover {
        span {
          color: ${colors.primary};
        }
        svg {
          fill: ${colors.primary};
        }
      }

      span {
        margin-left: 20px;
      }
    }
  }
`;

export const OverflowMenuNavigationItem = styled.div`
  padding: 1rem;
  color: ${colors.grey.normal};
  display: flex;
  align-items: center;
  transition: 1s ease all;
  font-size: 1.1rem;

  :hover {
    span {
      color: ${colors.primary};
    }
    svg {
      fill: ${colors.primary};
    }
  }

  span {
    margin-left: 20px;
  }
`;
