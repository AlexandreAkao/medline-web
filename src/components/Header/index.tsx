import { useNavigate } from 'react-router-dom';
import { useMediaQuery } from 'react-responsive';

import { Nav, LogoImg, NavItemsContainer } from 'components/Header/styles';
import MedlineLogo from 'assets/images/medline-logo.svg';
import ScreenSizes from 'styles/screenSizes';
import MenuBurger from 'components/MenuBurger';

function Header({
  icon = MedlineLogo,
  iconAlt = 'Medline Logo',
  isAuthenticated,
  children,
}: IHeaderProps) {
  const isTablet = useMediaQuery({ query: `(max-width: ${ScreenSizes.medium})` });
  const navigate = useNavigate();

  const handleClickIcon = () => {
    navigate('/');
  };

  return (
    <Nav isAuthenticated={isAuthenticated} data-testid="test-navigation">
      <LogoImg src={icon} alt={iconAlt} onClick={handleClickIcon} />
      {isTablet ? (
        <MenuBurger>{children}</MenuBurger>
      ) : (
        <NavItemsContainer data-testid="test-navigation-items">{children}</NavItemsContainer>
      )}
    </Nav>
  );
}

export default Header;
