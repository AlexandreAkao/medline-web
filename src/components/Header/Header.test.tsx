import { screen } from '@testing-library/react';
import { composeStories } from '@storybook/testing-react';
import userEvent from '@testing-library/user-event';
import { vi } from 'vitest';

import * as stories from 'components/Header/Header.stories';
import { renderWithRouter } from 'utils/testWrapper';
import Header from 'components/Header';

const { Authenticated, WithoutAuthenticated } = composeStories(stories);

const navigate = vi.fn();
vi.mock('react-router-dom', async () => {
  const reactRouterDom = await vi.importActual('react-router-dom');

  return {
    ...(reactRouterDom as Record<string, unknown>),
    useNavigate: () => navigate,
  };
});

describe('Header', () => {
  const renderComponent = (customProps: Partial<IHeaderProps> = {}) => {
    const props: IHeaderProps = {
      isAuthenticated: false,
      icon: 'https://cdn-icons-png.flaticon.com/512/5277/5277459.png',
      iconAlt: 'icon',
      children: undefined,
      ...customProps,
    };
    return renderWithRouter(<Header {...props} />);
  };

  it('should match with snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it('should render authenticated header', () => {
    renderWithRouter(<Authenticated />);
    const navigationContainer = screen.getByTestId('test-navigation');
    const itemsContainer = screen.getByTestId('test-navigation-items');
    expect(navigationContainer).toBeInTheDocument();
    expect(navigationContainer).toHaveStyle(`
      box-shadow: 0px 4px 10px 1px rgba(0,0,0,0.25)
    `);
    expect(itemsContainer.childElementCount).toBe(4);
  });

  it('should render header without authenticated', () => {
    renderWithRouter(<WithoutAuthenticated />);
    const navigationContainer = screen.getByTestId('test-navigation');
    const itemsContainer = screen.getByTestId('test-navigation-items');
    expect(navigationContainer).toBeInTheDocument();
    expect(navigationContainer).toHaveStyle(`
      box-shadow: ${''}
    `);
    expect(itemsContainer.childElementCount).toBe(2);
  });

  it('should render header without items', () => {
    renderComponent();
    const navigationContainer = screen.getByTestId('test-navigation');
    const itemsContainer = screen.getByTestId('test-navigation-items');
    expect(navigationContainer).toBeInTheDocument();
    expect(itemsContainer.childElementCount).toBe(0);
  });

  it('should redirect to LandingPage when clicked on icon', async () => {
    renderComponent();
    const iconElement = screen.getByRole('img');
    await userEvent.click(iconElement);

    expect(iconElement).toBeInTheDocument();
    expect(navigate).toBeCalledWith('/');
  });
});
