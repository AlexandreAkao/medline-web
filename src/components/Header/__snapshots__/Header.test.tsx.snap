// Vitest Snapshot v1

exports[`Header > should match with snapshot 1`] = `
<div>
  <header
    class="sc-gsnTZi jLFYRz"
    data-testid="test-navigation"
  >
    <img
      alt="icon"
      class="sc-dkzDqf dxiqfB"
      src="https://cdn-icons-png.flaticon.com/512/5277/5277459.png"
    />
    <nav
      class="sc-hKMtZM iKwNjC"
      data-testid="test-navigation-items"
    />
  </header>
</div>
`;
